hazelcast/src/main/java/com/hazelcast/sql/impl/worker/QueryFragmentExecutable.java

```diff
@@ -0,0 +1,218 @@
+            setupExecutor();
+
+            // Feed all batches to relevant inboxes first. Set the upper boundary on the number of batches to avoid
+            // starvation when batches arrive quicker than we are able to process them.
+            int maxOperationCount = operationCount.get();
+            int processedBatchCount = 0;
+
+            QueryOperation operation;
+
+            while ((operation = operations.pollFirst()) != null) {
+                if (operation instanceof QueryBatchExchangeOperation) {
+                    QueryBatchExchangeOperation operation0 = (QueryBatchExchangeOperation) operation;
+
+                    InboundHandler inbox = inboxes.get(operation0.getEdgeId());
+                    assert inbox != null;
+
+                    InboundBatch batch = new InboundBatch(
+                        operation0.getBatch(),
+                        operation0.isLast(),
+                        operation0.getCallerId()
+                    );
+
+                    inbox.onBatch(batch, operation0.getRemainingMemory());
+                } else {
+                    assert operation instanceof QueryFlowControlExchangeOperation;
+
+                    QueryFlowControlExchangeOperation operation0 = (QueryFlowControlExchangeOperation) operation;
+
+                    Map<UUID, OutboundHandler> edgeOutboxes = outboxes.get(operation0.getEdgeId());
+                    assert edgeOutboxes != null;
+
+                    OutboundHandler outbox = edgeOutboxes.get(operation.getCallerId());
+                    assert outbox != null;
+
+                    outbox.onFlowControl(operation0.getRemainingMemory());
+                }
+
+                if (++processedBatchCount >= maxOperationCount) {
+                    break;
+                }
+            }
+
+            operationCount.addAndGet(-1 * processedBatchCount);
+
+            IterationResult res = exec.advance();
+
+            // Send flow control messages if needed.
+            if (res != IterationResult.FETCHED_DONE) {
+                for (InboundHandler inbox : inboxes.values()) {
+                    inbox.sendFlowControl();
+                }
+            }
+
+            // If executor finished, notify the state.
+            if (res == IterationResult.FETCHED_DONE) {
+                completed = true;
+
+                stateCallback.onFragmentFinished();
+            }
+        } catch (Exception e) {
+            // Prevent subsequent invocations.
+            completed = true;
+
+            // Notify state about the exception to trigger cancel operation.
+            stateCallback.cancel(e);
+        }
+
+        // Unschedule the fragment with double-check for new batches.
+        unschedule();
+    }
+
+    @Override
+    public boolean schedule() {
+        // If the fragment is already scheduled, we do not need to do anything else, because executor will re-check queue state
+        // before exiting.
+        if (scheduled.get()) {
```

Comment:
It seems this check is not necessary because it is covered by the `scheduled.compareAndSet` logic.

Reply:
CAS is typically implemented as a write operation even if it doesn't change the value, causing contention on the bus. So here we employ `test-and-test-and-set` rather than `test-and-set` (CAS), to minimize the number of writes on a shared variable. 
This should be considered as a micro-optimization, and it is hard to prove that it adds value with some numbers. But in general, it is assumed that there should be fewer fragment execution attempts than batches, so IMO it should pay off.

Reply:
BTW, I slightly changed the reschedule algorithm, so that we reach this place even rarer.